pkg/controllers/machinesetsync/machineset_sync_controller.go

```diff
@@ -791,53 +770,401 @@ func (r *MachineSetSyncReconciler) createOrUpdateCAPIInfraMachineTemplate(ctx co
+
+			if condErr := r.applySynchronizedConditionWithPatch(
+				ctx, existingMAPIMachineSet, corev1.ConditionFalse, reasonFailedToUpdateMAPIMachineSet, updateErr.Error(), nil); condErr != nil {
+				return false, utilerrors.NewAggregate([]error{updateErr, condErr})
+			}
+
+			return false, updateErr
+		}
+
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// setChangedCAPIMachineSetStatusFields sets the updated fields in the CAPI machine set status.
+//
+//nolint:gocognit,funlen
+func setChangedCAPIMachineSetStatusFields(existingCAPIMachineSet, convertedCAPIMachineSet *clusterv1.MachineSet) {
+	// convertedCAPIMachineSet holds the computed and desired status changes, so apply them to the existing existingCAPIMachineSet.
+	existingCAPIMachineSet.Status.Replicas = convertedCAPIMachineSet.Status.Replicas
+	existingCAPIMachineSet.Status.ReadyReplicas = convertedCAPIMachineSet.Status.ReadyReplicas
+	existingCAPIMachineSet.Status.AvailableReplicas = convertedCAPIMachineSet.Status.AvailableReplicas
+	existingCAPIMachineSet.Status.FullyLabeledReplicas = convertedCAPIMachineSet.Status.FullyLabeledReplicas
+	existingCAPIMachineSet.Status.FailureReason = convertedCAPIMachineSet.Status.FailureReason
+	existingCAPIMachineSet.Status.FailureMessage = convertedCAPIMachineSet.Status.FailureMessage
+
+	// Update the conditions if they exist (ignoring lasttransitiontime), append new ones.
+	for i := range convertedCAPIMachineSet.Status.Conditions { //nolint:dupl
+		hasCondition := false
+
+		for j := range existingCAPIMachineSet.Status.Conditions {
+			if existingCAPIMachineSet.Status.Conditions[j].Type == convertedCAPIMachineSet.Status.Conditions[i].Type {
+				hasChanged := false
+				hasCondition = true
+
+				if existingCAPIMachineSet.Status.Conditions[j].Status != convertedCAPIMachineSet.Status.Conditions[i].Status {
+					existingCAPIMachineSet.Status.Conditions[j].Status = convertedCAPIMachineSet.Status.Conditions[i].Status
+					hasChanged = true
+				}
+
+				if existingCAPIMachineSet.Status.Conditions[j].Reason != convertedCAPIMachineSet.Status.Conditions[i].Reason {
+					existingCAPIMachineSet.Status.Conditions[j].Reason = convertedCAPIMachineSet.Status.Conditions[i].Reason
+					hasChanged = true
+				}
+
+				if existingCAPIMachineSet.Status.Conditions[j].Message != convertedCAPIMachineSet.Status.Conditions[i].Message {
+					existingCAPIMachineSet.Status.Conditions[j].Message = convertedCAPIMachineSet.Status.Conditions[i].Message
+					hasChanged = true
+				}
+
+				if hasChanged {
+					// Set the last transition time to the new condition's last transition time only if the condition has changed.
+					existingCAPIMachineSet.Status.Conditions[j].LastTransitionTime = convertedCAPIMachineSet.Status.Conditions[i].LastTransitionTime
+				}
+
+				break
+			}
+		}
+
+		if !hasCondition {
+			existingCAPIMachineSet.Status.Conditions = append(existingCAPIMachineSet.Status.Conditions, convertedCAPIMachineSet.Status.Conditions[i])
+		}
```

Comment:
Consider helper functions for this? I'm sure there are already helpers in places like CAPI we could take inspiration from?